libsnark/zk_proof_systems/plonk/plonk.tcc

```diff
@@ -0,0 +1,412 @@
+        omega[base_k1].begin(), omega[base_k1].end(), back_inserter(H_gen));
+    std::copy(
+        omega[base_k2].begin(), omega[base_k2].end(), back_inserter(H_gen));
+}
+
+/// permute the multiplicative subgroup H according to the wire
+/// permutation: (see [GWC19] Sect. 8), \see
+/// plonk_compute_roots_of_unity_omega, \see
+/// plonk_roots_of_unity_omega_to_subgroup_H
+template<typename FieldT>
+void plonk_permute_subgroup_H(
+    const std::vector<FieldT> &H_gen,
+    const std::vector<size_t> &wire_permutation,
+    std::vector<FieldT> &H_gen_permute)
+{
+    assert(H_gen.size() > 0);
+    for (size_t i = 0; i < H_gen.size(); ++i) {
+        H_gen_permute[i] = H_gen[wire_permutation[i] - 1];
+    }
+}
+
+/// compute the permutation polynomials S_sigma_1, S_sigma_2,
+/// S_sigma_2 (see [GWC19], Sect. 8.1)
+template<typename FieldT>
+void plonk_compute_permutation_polynomials(
+    const std::vector<FieldT> &H_gen_permute,
+    const size_t num_gates,
+    std::vector<polynomial<FieldT>> &S_polys)
+{
+    assert(S_polys.size() == NUM_HSETS);
+    assert(S_polys[0].size() == num_gates);
+    assert(H_gen_permute.size() == (NUM_HSETS * num_gates));
+    for (size_t i = 0; i < NUM_HSETS; ++i) {
+        typename std::vector<FieldT>::const_iterator begin =
+            H_gen_permute.begin() + (i * num_gates);
+        typename std::vector<FieldT>::const_iterator end =
+            H_gen_permute.begin() + (i * num_gates) + (num_gates);
+        std::vector<FieldT> S_points(begin, end);
+        plonk_interpolate_polynomial_from_points<FieldT>(S_points, S_polys[i]);
+    }
+}
+
+// A wrapper for multi_exp_method_BDLO12_signed() dot-product a
+// vector of group elements in G1 (curve points) with a vector of
+// scalar elements in Fr
+template<typename ppT>
+libff::G1<ppT> plonk_multi_exp_G1(
+    const std::vector<libff::G1<ppT>> &curve_points,
+    const std::vector<libff::Fr<ppT>> &scalar_elements)
+{
+    assert(curve_points.size() == scalar_elements.size());
+    const size_t chunks = 1;
+    libff::G1<ppT> product = libff::multi_exp<
+        libff::G1<ppT>,
+        libff::Fr<ppT>,
+        libff::multi_exp_method_BDLO12_signed>(
+        curve_points.begin(),
+        curve_points.end(),
+        scalar_elements.begin(),
+        scalar_elements.end(),
+        chunks);
+    return product;
+}
+
+/// Evaluate a polynomial F at the encrypted secret input
+/// \secret^i*G_1 ie. compute f(\secret)*G1 = [f(\secret)]_i
+///
+/// INPUT
+///
+/// \param[in] secret_powers_g1: \secret^i*G1:
+///            0\le{i}<max_degree(Q[j]): 0\le{j}<n f_poly: a
+///            polynomial
+///
+/// OUTPUT
+///
+/// \param[out] [f_poly(\secret)]_1, 0\le 1<n : the "encrypted"
+///             evaluation of the polynomial f_poly in the secret
+///             parameter \secret (the toxic waste) multiplied by the
+///             group generator G_1 i.e. compute f_poly(\secret)*G_1
+template<typename ppT>
+libff::G1<ppT> plonk_evaluate_poly_at_secret_G1(
```

Comment:
Reminder to discuss whether we can use the BGG17 trick "FFT in the coefficients" here, by precomputing $[ L_i( \tau ) ]_1$
for secret  $\tau$  and Lagrange polynomials $L_i$.